### PR TITLE
[compiler-v2] Fix #13744 by tracking which assignments come from var defs in stackless_bytecode

### LINKED ISSUE #13744: [Bug] [Compiler-V2] Duplicate error messages from unused varaibles and unused assignments
# 🐛 Bug

There are duplicated warnings from unused variable checker (implemented in `flow_insensitive_checker`) and unused assignment checker (implemented in `unused_assignment_checker`). 

## To reproduce

See for example,

https://github.com/aptos-labs/aptos-core/blob/d76d5bfa6ec2d259b1e83674a7b05e6177f9a53f/third_party/move/move-compiler-v2/tests/unused-assignment/v1-locals/unused_resource.exp

## Expected Behavior

There should only be one error message for one location.

### LINKED ISSUE #13880: [Bug][move-compiler-v2] Spurious Unused assignment warnings with inlining in certain tests
# 🐛 Bug

I've only seen these messages in some big runs, where V2 is used to pre-compile libraries for aptos-transactional-test-harness (for V2 tests).

```
% UB=1 cargo test --profile ci -p move-compiler -p move-compiler-v2 -p move-compiler-v2-transactional-tests -p move-compiler-transactional-tests -p aptos-transactional-test-harness |& tee test.out2
...

test runner::aptos_test_harness/randomness_safety.move                 ... ok
warning: Unused assignment to `to`. Consider removing or prefixing with an underscore: `_to`
    ┌─ /Users/brm/code/aptos-core/aptos-move/framework/aptos-framework/sources/object.move:544:16
    │
544 │     inline fun transfer_raw_inner(object: address, to: address) acquires ObjectCore {
    │                ^^^^^^^^^^^^^^^^^^
    ·
621 │         transfer_raw_inner(object_addr, BURN_ADDRESS);
    │         --------------------------------------------- from a call inlined at this callsite

warning: Unused assignment to `c`. Consider removing or prefixing with an underscore: `_c`
     ┌─ /Users/brm/code/aptos-core/aptos-move/framework/aptos-stdlib/sources/math64.move:41:23
     │
  41 │     public inline fun mul_div(a: u64, b: u64, c: u64): u64 {
     │                       ^^^^^^^
     │
     ┌─ /Users/brm/code/aptos-core/aptos-move/framework/aptos-framework/sources/delegation_pool.move:2000:13
     │
2000 │             math64::mul_div(active - pool_active, pool.operator_commission_percentage, MAX_FEE)
     │             ----------------------------------------------------------------------------------- from a call inlined at this callsite

warning: Unused assignment to `c`. Consider removing or prefixing with an underscore: `_c`
     ┌─ /Users/brm/code/aptos-core/aptos-move/framework/aptos-stdlib/sources/math64.move:41:23
     │
  41 │     public inline fun mul_div(a: u64, b: u64, c: u64): u64 {
     │                       ^^^^^^^
     │
     ┌─ /Users/brm/code/aptos-core/aptos-move/framework/aptos-framework/sources/delegation_pool.move:2008:13
     │  
2008 │ ╭             math64::mul_div(
2009 │ │                 pending_inactive - pool_pending_inactive,
2010 │ │                 pool.operator_commission_percentage,
2011 │ │                 MAX_FEE
2012 │ │             )
     │ ╰─────────────' from a call inlined at this callsite

```
The code currently looks like:
```
module aptos_framework::object {
   ...
    inline fun transfer_raw_inner(object: address, to: address) acquires ObjectCore {
        let object_core = borrow_global_mut<ObjectCore>(object);
        if (object_core.owner != to) {
            if (std::features::module_event_migration_enabled()) {
                event::emit(
                    Transfer {
                        object,
                        from: object_core.owner,
                        to,
                    },
                );
            };
            event::emit_event(
                &mut object_core.transfer_events,
                TransferEvent {
                    object,
                    from: object_core.owner,
                    to,
                },
            );
            object_core.owner = to;
        };
    }
    ...
    public entry fun burn<T: key>(owner: &signer, object: Object<T>) acquires ObjectCore {
        let original_owner = signer::address_of(owner);
        assert!(is_owner(object, original_owner), error::permission_denied(ENOT_OBJECT_OWNER));
        let object_addr = object.inner;
        move_to(&create_signer(object_addr), TombStone { original_owner });
        transfer_raw_inner(object_addr, BURN_ADDRESS);
    }
    ...

```

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Add a field to `Bytecode::Assign` to track `DefOrAssign` for each
assignment, so that assignments that correspond to variable
definitions can be rememebered and the subsequent "unused assignment"
pass can avoid warning again for them.

This required removing a previous (premature?) optimization which
tried to avoid a temp in the case of a var assignment.

Examining output code suggests that the temp is almost always removed
by copy propagation, but I see a few examples where the redundant copy
is not removed, see
- `aptos-core/third_party/move/move-compiler-v2/tests/copy-propagation/immut_refs_2.exp`
Perhaps we are not removing reference valued temps?

In any case, removing that optimization incidentally improves the
remaining "unused assignment" warnings, by better placing the warning
on the variable whose assignment is unused, rather than the assignment
RHS.

I found it useful to flag bytecode "def" versus assignment in the
output, but it leads to too big a diff (too many tests outputs include
intermediate bytecode) so I removed that functionality for now.

Fixes #13744 .
Fixes #13880.

## How Has This Been Tested?

So far, just basic regression tests for compilers and users.

## Key Areas to Review

Somehow this also some incorrect "unused assignment" warnings for
move-prover tests, though I don't know how.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [X] Other (specify)
   Move compiler

## Checklist
- [X] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I identified and added all stakeholders and component owners affected by this change as reviewers
- [X] I tested both happy and unhappy path of the functionality
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/src/bytecode_generator.rs [hunk 1]
```diff
@@ -367,15 +367,7 @@ impl<'env> Generator<'env> {
...
-                    if let Pattern::Var(var_id, sym) = pat {
-                        // For the common case `let x = binding; ...` avoid introducing a
-                        // temporary for `binding` and directly pass the temp for `x` into
-                        // translation.
-                        let local = self.find_local_for_pattern(*var_id, *sym, Some(&scope));
-                        self.without_reference_mode(|s| s.gen(vec![local], binding))
-                    } else {
-                        self.gen_assign(pat.node_id(), pat, binding, Some(&scope));
-                    }
+                    self.gen_assign(pat.node_id(), pat, binding, Some(&scope));
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/src/bytecode_generator.rs:370]: Could you explain on the change here?

**Review by wrwg — COMMENTED** [review]
This is a very intrusive change for a minor bug. (I'm not sure why this is marked as urgent).

- I have never seen any IR with the need to distinguish between assign or definition. 
- Every future code working and rewriting stackless bytecode now has to deal with the complexity of getting this new field value right. (The size of the PR is a first glimpse of this.)
- If we want to avoid duplicate warnings, there should be other ways then redefining the IR 

Please let us discuss extensions like this to the IR first. Maybe there are reasons we have to do this,  but I'm not seeing them

**Review by wrwg — CHANGES_REQUESTED** [review]
See comment

**Comment by brmataptos** [comment]
> This is a very intrusive change for a minor bug. (I'm not sure why this is marked as urgent).
> 
> * I have never seen any IR with the need to distinguish between assign or definition.
> * Every future code working and rewriting stackless bytecode now has to deal with the complexity of getting this new field value right. (The size of the PR is a first glimpse of this.)
> * If we want to avoid duplicate warnings, there should be other ways then redefining the IR
> 
> Please let us discuss extensions like this to the IR first. Maybe there are reasons we have to do this, but I'm not seeing them

(1) I've never seen a compiler where we wanted to implement warnings so late in the game.  I tried to push to do more in the AST, but we wound up with the unused assignment warning analysis implemented on the stackless bytecode.  This close to release, implementing a dataflow analysis on the AST is much riskier than a quick fix like this.  It's easy enough to undo later.

(2) This field is no more intrusive than the `AssignKind` field; in most cases, it can be ignored.

(3) This is urgent because my PR (#14281) is blocked from landing due to inconsistent test outputs on opt versus non-opt code paths for transactional tests.  If you would rather fix your test framework to allow distinct outputs for opt versus non-opt, feel free.  Though I don't think that's a great idea.

This also fixes #13880, which was already Urgent.

**Pull request closed without merging**